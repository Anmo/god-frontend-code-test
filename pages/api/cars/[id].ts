import type { NextApiRequest, NextApiResponse } from "next";

import cars from "../../../public/api/cars.json";
import type { Car } from "../../../src/types";

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Car | { error: string }>
) {
  const car = cars.find(({ id }) => id === req.query.id);

  if (!car) {
    return res.status(404).json({ error: "Car not found" });
  }

  res.status(200).json(car);
}
