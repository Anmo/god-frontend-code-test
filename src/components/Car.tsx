import React from "react";
import Image from "next/image";
import Link from "next/link";
import { Icon } from "@volvo-cars/react-icons";

import type { Car } from "../types";

export const Card: React.FC<
  Car & { withBackButton?: boolean; withBuyButton?: boolean }
> = ({
  id,
  bodyType,
  modelName,
  modelType,
  imageUrl,
  withBackButton,
  withBuyButton,
}) => {
  return (
    <div className="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start">
      <p className="font-medium">{bodyType}</p>
      <div className="flex-col md:flex-row gap-8">
        <p className="font-medium">{modelName}</p>
        <p className="font-medium text-secondary">{modelType}</p>
      </div>
      <Image
        alt={modelName}
        className="cover aspect-4/3 w-full mb-8"
        height="300"
        src={imageUrl}
        style={{
          height: "auto",
        }}
        width="400"
      />
      <div className="flex flex-wrap gap-x-24 justify-center">
        {withBackButton || withBuyButton ? (
          <Link className="text-accent-blue" href="/">
            <Icon type="navigation-chevronback-16" color="currentColor" /> Back
          </Link>
        ) : (
          <Link className="text-accent-blue" href={`/learn/${id}`}>
            Learn{" "}
            <Icon type="navigation-chevronforward-16" color="currentColor" />
          </Link>
        )}
        {withBuyButton ? (
          <button className="text-accent-blue" disabled>
            BUY
          </button>
        ) : (
          <Link className="text-accent-blue" href={`/shop/${id}`}>
            Shop{" "}
            <Icon type="navigation-chevronforward-16" color="currentColor" />
          </Link>
        )}
      </div>
    </div>
  );
};
