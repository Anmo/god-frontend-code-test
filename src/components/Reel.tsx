import React, { useRef } from "react";
import { IconButton } from "@volvo-cars/react-icons";
import { useReel } from "@volvo-cars/react-headless";

export const Reel: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const ref = useRef<HTMLDivElement>(null);
  const { activeIndex, indicatorCount, previousButtonProps, nextButtonProps } =
    useReel({ ref });

  return (
    <>
      <div
        ref={ref}
        className="container-max reel scrollbar-none gap-x-gutter py-16 px-pagemargin lg:px-0"
      >
        {children}
      </div>
      <div className="reel-indicators md:hidden" aria-hidden>
        {[...Array(indicatorCount).keys()].map((_, index) => (
          <div key={index} aria-current={activeIndex === index} />
        ))}
      </div>
      <div className="button-group justify-end mr-16 mt-16 until-md:hidden">
        <IconButton
          iconName="navigation-chevronback"
          {...previousButtonProps}
        />
        <IconButton iconName="navigation-chevronforward" {...nextButtonProps} />
      </div>
    </>
  );
};
