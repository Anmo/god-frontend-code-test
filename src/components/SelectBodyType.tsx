import useSWR from "swr";
import { Select } from "@volvo-cars/react-forms";

import fetcher from "../utils/fetcher";

import type { CarBodyType } from "../types";
import { useCallback } from "react";

export const SelectBodyType: React.FC<{
  className?: string;
  onChange: (bodyType: CarBodyType | undefined) => void;
}> = ({ className, onChange }) => {
  const { data } = useSWR<CarBodyType[]>("/api/bodyTypes", fetcher);

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      onChange(e.target.value || undefined);
    },
    [onChange]
  );

  return (
    <Select
      className={className}
      name="bodyType"
      label="Select body type"
      onChange={handleChange}
      defaultValue={""}
    >
      <option value="">All</option>
      {data?.map((bodyType) => (
        <option key={bodyType} value={bodyType}>
          {bodyType}
        </option>
      ))}
    </Select>
  );
};
