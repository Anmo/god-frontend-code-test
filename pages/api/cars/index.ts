import type { NextApiRequest, NextApiResponse } from "next";

import cars from "../../../public/api/cars.json";
import type { CarBodyType, Cars } from "../../../src/types";

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Cars>
) {
  const bodyType = req.query.bodyType as CarBodyType | undefined;
  return res
    .status(200)
    .json(bodyType ? cars.filter((car) => car.bodyType === bodyType) : cars);
}
