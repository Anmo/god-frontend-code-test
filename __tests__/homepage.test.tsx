import { ReactNode } from "react";
import { vi, describe, expect, it, beforeEach, Mock, afterEach } from "vitest";
import { render, screen } from "@testing-library/react";
import useSWR from "swr";

import { SelectBodyType } from "../src/components/SelectBodyType";
import cars from "../public/api/cars.json";
import Homepage from "../pages";

vi.mock("swr");
vi.mock("../src/components/SelectBodyType");
vi.mock("../src/components/Reel", () => ({
  Reel: ({ children }: { children: ReactNode }) => <>{children}</>,
}));
const useSWRMock = useSWR as Mock;
const SelectBodyTypeMock = SelectBodyType as Mock;

describe("Homepage", () => {
  let onChange: (bodyType: string | undefined) => void;

  beforeEach(() => {
    SelectBodyTypeMock.mockImplementation(
      ({
        onChange: _onChange,
        ...props
      }: { onChnage: (bodyType: string) => void } & Record<string, any>) => {
        onChange = _onChange;

        return <div {...props}>SelectBodyType</div>;
      }
    );
    useSWRMock.mockReturnValue({});
  });

  afterEach(() => {
    useSWRMock.mockClear();
  });

  it("renders loading state", () => {
    // When
    render(<Homepage />);

    // Then
    expect(screen.getByText("Loading...")).toBeDefined();
  });

  it("renders error state", () => {
    // Given
    useSWRMock.mockReturnValue({ error: "Something wrong is not correct!" });

    // When
    render(<Homepage />);

    // Then
    expect(screen.getByText("Failed to load")).toBeDefined();
  });

  it("renders error state", () => {
    // Given
    useSWRMock.mockReturnValueOnce({ data: cars });

    // When
    const { container } = render(<Homepage />);

    // Then
    expect(container).matchSnapshot();
  });

  it("fetches filtered cars", async () => {
    // Given
    render(<Homepage />);

    // When
    onChange("suv");
    await new Promise((resolve) => setTimeout(resolve)); // next tick

    // Then
    expect(useSWRMock).toBeCalledWith(
      "/api/cars?bodyType=suv",
      expect.any(Function)
    );
  });
});
