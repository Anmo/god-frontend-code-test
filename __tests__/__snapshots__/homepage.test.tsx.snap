// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homepage > renders error state 1`] = `
<div>
  <div
    class="flex-col md:flex-row gap-8 justify-between items-center px-pagemargin py-8"
  >
    <img
      class="wordmark text-primary"
      height="8"
      src="https://www.volvocars.com/static/shared/images/volvo-spread-wordmark.svg"
      style="--mask-url: url('https://www.volvocars.com/static/shared/images/volvo-spread-wordmark.svg');"
      width="104"
    />
    <div
      class="w-xs"
    >
      SelectBodyType
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      suv
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        XC90 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="XC90 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fxc90_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/xc90-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/xc90-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      suv
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        XC60 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="XC60 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fxc60_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fxc60_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fxc60_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/xc60-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/xc60-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      suv
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        XC40 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="XC40 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fxc40_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fxc40_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fxc40_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/xc40-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/xc40-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      suv
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        XC40 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        pure electric
      </p>
    </div>
    <img
      alt="XC40 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fxc40_bev.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fxc40_bev.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fxc40_bev.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/xc40-bev"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/xc40-bev"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      estate
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        V90 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="V90 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fv90_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fv90_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fv90_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/v90-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/v90-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      estate
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        V60 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="V60 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fv60_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fv60_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fv60_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/v60-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/v60-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      sedan
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        S90 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="S90 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fs90_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fs90_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fs90_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/s90-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/s90-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
  <div
    class="flex-col flex-shrink-0 flex-grow-0 gap-y-8 w-xs h-full snap-start"
  >
    <p
      class="font-medium"
    >
      sedan
    </p>
    <div
      class="flex-col md:flex-row gap-8"
    >
      <p
        class="font-medium"
      >
        S60 Recharge
      </p>
      <p
        class="font-medium text-secondary"
      >
        plug-in hybrid
      </p>
    </div>
    <img
      alt="S60 Recharge"
      class="cover aspect-4/3 w-full mb-8"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fs60_recharge.jpg&w=828&q=75"
      srcset="/_next/image?url=%2Fimages%2Fs60_recharge.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2Fs60_recharge.jpg&w=828&q=75 2x"
      style="color: transparent; height: auto;"
      width="400"
    />
    <div
      class="flex flex-wrap gap-x-24 justify-center"
    >
      <a
        class="text-accent-blue"
        href="/learn/s60-recharge"
      >
        Learn
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
      <a
        class="text-accent-blue"
        href="/shop/s60-recharge"
      >
        Shop
         
        <img
          class="icon-mask"
          height="16"
          role="presentation"
          src="https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg"
          style="--icon-url: url(https://www.volvocars.com/static/shared/icons/v2/navigation-chevronforward-16.svg);"
          width="16"
        />
      </a>
    </div>
  </div>
</div>
`;
